torch/csrc/generic/Tensor.cpp

```diff
@@ -496,14 +503,354 @@ static bool THPTensor_(_indexOnce)(PyObject *index, int &indexed_dim,
+                                               THTensor_(storage)(LIBRARY_STATE indexed.get())),
+                                           1);
+
+  return true;
+}
 
+// Should called, written in such a way that if any of the parameters are not
+// initialized we still don't crash
+static void THPTensor_(_advancedIndexCommonCleanup)(
+    std::vector<THIndexTensor*>& broadcasted,
+    THIndexTensor *linearIndices,
+    THTensor *flattened) {
+
+  // Cleanup all TH memory
+  for (const auto& bTensor : broadcasted) {
+    THIndexTensor_(free)(LIBRARY_STATE bTensor);
+  }
+  if (linearIndices) THIndexTensor_(free)(LIBRARY_STATE linearIndices);
+  if (flattened) THTensor_(free)(LIBRARY_STATE flattened);
+}
+
+static bool THPTensor_(_advancedIndexGet)(PyObject *index, THTensorPtr &tresult)
+{
+  std::vector<THIndexTensor*> broadcasted;
+  THIndexTensor *linearIndices = NULL;
+  THTensor *flattened = NULL;
+  bool success = THPTensor_(_advancedIndexCommonInit)(
+      index, tresult, broadcasted, &linearIndices, &flattened);
+
+  if (success) {
+    THTensor *result = THTensor_(new)(LIBRARY_STATE_NOARGS);
+
+    // Index Select makes a copy of the storage, thus it is enforcing NumPy semantics, which says that the
+    // array returned by advanced indexing is a copy, not a view
+    THTensor_(indexSelect)(LIBRARY_STATE result, flattened, 0, linearIndices);
+
+    // In the event that the indexing Tensors are not vectors, we need to reshape
+    // the result to be the appropriate shape. Note that we do not use the strides
+    // of a broadcast Tensor, as they may have been mucked with to support the
+    // broadcast semantics
+    THTensor_(resizeNd)(LIBRARY_STATE result,
+                        THIndexTensor_(nDimension)(LIBRARY_STATE broadcasted.at(0)),
+                        broadcasted.at(0)->size,
+                        NULL);
+
+    // result ptr takes ownership of result tensor, and implicitly frees the
+    // indexed one
+    tresult = result;
+  }
+
+  THPTensor_(_advancedIndexCommonCleanup)(broadcasted, linearIndices, flattened);
+  return success;
+}
+
+static bool THPTensor_(_advancedIndexSet)(PyObject *index, THTensorPtr &dest, PyObject *src)
+{
+
+  std::vector<THIndexTensor*> broadcasted;
+  THIndexTensor *linearIndices = NULL;
+  THTensor *flattened = NULL;
+  bool success = THPTensor_(_advancedIndexCommonInit)(
+      index, dest, broadcasted, &linearIndices, &flattened);
+
+  if (success) {
+    if (THPUtils_(checkReal)(src)) {
+      real v = THPUtils_(unpackReal)(src);
+      THTensor_(indexFill)(LIBRARY_STATE flattened, 0, linearIndices, v);
+    } else if (THPTensor_(Check)(src)) {
+      // Because we are doing an index copy, we need to make sure of two things:
+      // 1. the src Tensor is 1D and
+      // 2. the src is made contiguous before being flattened into a 1D view, if
+      // necessary
+
+      THTensor *contiguous;
+      if (THTensor_(isContiguous(LIBRARY_STATE ((THPTensor*)src)->cdata))) {
```

Comment:
You can just call `THTensor_(newContiguous)`, as [it internally checks if the tensor is contiguous](https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensor.c#L190-L199) before performing the memory copy (skipping if it is contiguous).